### ISSUE TITLE
Questions about Dynd type system

### ISSUE THREAD

**Issue by datnamer** [libdynd/libdynd#985]
Hi Everyone, I've been following this discussion in #979 and I have some questions about Dynd's type system.

Hope its ok to open an issue for this, wasn't sure where else to post. 

I don't know much C++, but I was a bit confused by the muliple ways of describing data that Dynd seems to have.

There seems to be datashape/metadata, symbolic types, Concrete Types (analogous to class seems like), properties (ie T.real ), and @insertinterestingnamehere discussed including traits (I think this is great to faciliate lightweight abstractions for code reuse). 

How do these all fit together and what is the separation of duties? 

Also, Will we be able to define a Type so that its fields are typed as whatever data it is parameterized with? (like parametric types).

**Comment by mwiebe** [comment]
Hey @datnamer this is a good place to start such a discussion, the mailing list https://groups.google.com/forum/#!forum/libdynd-dev is also good, we haven't been utilizing that enough! I'll try to answer and give some insight.

Datashape is the name for the grammar used for string representations of DyND types. This is what "DyND uses datashape" means, that DyND has a parser for the datashape grammar and prints its types as datashape.

The basic idea for concrete representation of data is that the data exists in memory, with a memory layout of the style static languages typically have, but described with a dynamic type object. Concrete types are the types able to perform this duty.

Some operations, most notably slicing, are fundamental to the NumPy-style array programming DyND is born out of. The DyND metadata is a mechanism for supporting the strided array technique NumPy uses without hardcoding that as the only additional memory layout information. Each DyND type uses a fixed amount of memory that allows slicing of arrays, or field subsetting of structures in a way that produces views without modifying the data. Thus to describe memory a DyND type, corresponding metadata, and raw memory pointer are required and sufficient.

The properties are like properties in most other languages. We've taken a cue from the unified call syntax of D and other languages where t.f() can resolve to calling f(t) if such an overload exists. The details of that aren't fully worked out, the basics are in place but this will likely keep evolving as we gain experience with the approach's pros and cons.

The discussion in #979 is about dynamic properties exposed by the type objects themselves. For example a structure type contains arrays of field names and field types. It involves restricting them somewhat to allow a cleaner high level structure of the DyND library that isolates the ndt::type component from the nd::array and nd::callable parts.

Symbolic types are for describing types of data where some parts of the type aren't known. For example a 2D array "Fixed * Fixed * Scalar", or a 3D float array "Fixed**3 * float32". There is some planned type manipulation functionality to help with work with these, for example substituting in an array shape, substituting named type variables with another type, doing search/replace of type subcomponents. Say you have an array of type "3 * 4 * float32" and you want an equivalently-shaped array of bools. Something like tp.replace("Scalar", "bool") should be implemented to allow substitution of the data type without affecting the dimensions. These symbolic types are also used for describing the nd::callable signatures.

For your last question, could you maybe give an example of what you'd want to do with that? I can think of a few possible interpretations of what you said, and a specific example would be much better than me trying to speculate.

**Comment by datnamer** [comment]
All that definitely makes sense, thanks for taking the time to answer my questions. 

Is there a way to overload functions on some capability or interface of a type rather than a concrete representation of one?  IE  method that dispatches on any types that implement methods or fields x and y? Or would this come in with multiple inheritance?

Regarding my last questions on the previous post, I'll take an excerpt  out of the Julia manual, which explains it clearer than I can :

> Parametric  types can take parameters, so that type declarations actually introduce a whole family of new types — one for each possible combination of parameter values. 

> Type parameters are introduced immediately after the type name, surrounded by curly braces:
> 
> type Point{T}
>   x::T
>   y::T
> end
> This declaration defines a new parametric type, Point{T}, holding two “coordinates” of type T. What, one may ask, is T? Well, that’s precisely the point of parametric types: it can be any type at all (or a value of any bits type, actually, although here it’s clearly used as a type). Point{Float64} is a concrete type equivalent to the type defined by replacing T in the definition of Point with Float64

**Comment by insertinterestingnamehere** [comment]
@mwiebe Thanks for the awesome write up.

@datnamer With regards to type traits, I was originally referring to having something similar to the classifications in http://en.cppreference.com/w/cpp/header/type_traits for our own scalar types in C++. While adding some equivalent to our dynamic type system is possible, we haven't really discussed that much yet.

With regards to templated types, we do have something like this, but the interface is different from the templating systems you're referring to. One example is our option type which is a parameterized type that can wrap any given type as a child (though it's only implemented for builtin types right now). Most of our callables (function-like objects) are implemented using dynamic registries that dispatch on the types of the arrays given in a call. For a set of types to work with a given callable, an overload needs to be added to the callable that matches that particular set of types. I'd love to see further simplifications to this process and have several ideas about how to simplify things in C++, but that's what we have thus far.

Our parameterized types currently work by getting their own specialization when we dispatch in a callable call where they can do whatever they need to (like determine if a value is missing or not) and then call other type-specific implementations of things as needed. Any type-generic interface would rely on the parameter types having the needed overloads provided for them in the callables used to perform a given operation. That's our current equivalent of a templating mechanism.

Since all our types are dynamic, you can't simply make a C++ template and then transform it into a parameterized type that provides a given interface. This can be done for types that are implemented as C++ classes and are known at compile time, but not for dynamically added types. The compile-time nature of C++ templates makes that impossible. That said, the correct overloads for a given callable can be added dynamically.

One of the main purposes of what we're doing here is to provide a uniform ABI that is usable across different languages and runtimes. In that context, template-ized syntaxes are language specific. If we build interfaces to interact with C++ templates or type systems in some other JIT based environment (numba, Julia, etc.), we will have to rely on the templating mechanisms of that particular language and build ways to interact with our dynamic registries in some sensible way. In C++ this means providing interfaces to instantiate templates with various combinations of types and register the resulting overloads in a callable. In other languages that would mean something different. We already have some basic numba interoperability built into our Python bindings that forwards calls with a given set of types into numba's jit compiler rather than requiring that a given specialization be compiled in advance. I could see us adding things like that in other JIT based environments, but we haven't done it yet. I really like the idea, but I have no idea if/when that will happen or what the interface might look like.

As far as inheritance goes, we're in the process of working that in to our type dispatching system right now. Currently we dispatch based on strict type matches. Some generalizations there are needed, but the matter isn't fully settled either. @izaid has spent a little time looking into how to follow inheritance structures in our dispatch system, but it's not finished yet. Another related topic are implicit conversions. I put together a design doc on that a while ago, but have been dealing with other issues since then. Both of these are things we've looking at and would like to get finished. No final decisions or implementations are ready though.

In summary, the core functionality and type system is based primarily on registries right now. The way the registries are set up is generic enough to accomodate parameterized types and generic programming in most senses. That said, templating mechanisms require either forwarding calls back through our dynamic registries to get the right specializations for particular operations (as we've been doing in C++) or forwarding calls to a JIT compiler that can provide low-level type-specific functions on the fly (as with our numba interop).

**Comment by datnamer** [comment]
@insertinterestingnamehere  Thanks very much for the details. I'm going to read and digest what you wrote (some of the terminology is unfamiliar to me coming from a python+some julia background.) 

In the meantime, I just wanted to post this twitter conversation:  https://twitter.com/johnmyleswhite/status/701794065361162240

The Julia folks have spent many many people hours iterating, writing and deploying their multiple dispatch, abstract inheriatnace type system and I think it would behoove Dynd to draw on this experience.

**Comment by mwiebe** [comment]
I definitely agree with you on learning from Julia's experience. Based on all the Julia reading I've done, it looks very similar in substance to programming with C++ templates, with the twist about dynamic JIT and the ability for the results of a dispatch to change at run-time when new signatures are added. (There's a paper http://www.stroustrup.com/multimethods.pdf which proposes a C++ feature which might get a similar effect.)

Within C++, there's been a lot of work on how to formalize the 'concept' which was the basis of the STL and other libraries written in a generic fashion. (It's interesting to read Stepanov's account of of his generic programming research, my recollection is that lisp was the first language the STL was written in.) The comparisons between the C++ concept proposals and Haskell typeclasses are also worthwhile reading to understand the differences in goals and design choices going on there.

One big difference between DyND and Julia is that we're trying to keep the scope of DyND within library territory and avoid getting too far into language design. This means we're constraining ourselves to using idioms which map well to a C++ binding and a Python binding, and some of the fundamental performance characteristics are different. C++ being a static language and Python being a dynamic language mean they interact differently with the library. Adding Numba or another JIT into the mix gives some behaviours more comparable to Julia.

**Comment by datnamer** [comment]
Interesting.

I understand there are  technical limitations of python semantics and syntax. However, knowing what I know about Julia syntax/semantics and the existence of things like blaze(with symbolic manipulation), numba, decorators and the fact that datashape is already doing string parsing, are  these limitations are fundamentally insurmountable?

There is the issue of python idioms, but I don't know how much being conservative to maintain these is good for the long term future of the python ecosystem. (there is also numpy) 

An example is traits vs inheritance. JMW bring this up for issues of method dispatch ambiguity (and I'm sure other nuances I don't understand), but there is another side to this. Being able to add interfaces onto types that are already declared, to me seems like a perfect sort of disciplined, typed monkeypatching. 

This would allow the ecosystem to grow  with small compo-sable and interacting  packages that simply extend other's types without the messiness of subclassing concretely typed objects with multimethods.

What about using blaze for metaprogramming to compose Dynd func at a higher level of abstraction? Here is a excerpt I found  from the mailing list:

> Actually, what I'm hoping for is something like Terra (http://terralang.org/) or Halide(http://halide-lang.org/) but within the python ecosystem - using data descriptors (dynd), computation descriptors (blaze) and templated computation kernels (numba) to create a fused, runtime optimized kernel.
> 
> Stanley - have you seen the work on Lightweight Modular Staging (in either Scala - https://scala-lms.github.io/ or Lua - http://terralang.org/)

Some other interesting ideas can be found in flypy: https://github.com/flypy/flypy  


I've also heard complaints that Julia is still too loose with its typing and scientists are having trouble understanding where types suddenly change affecting performance. I think this is an area that blaze is already ahead with datashape and return types. 

Anyway, just some thoughts of mine- feel free to take with a grain of salt since I haven't contributed to Dynd and am just watching eagerly from the sidelines :)

**Comment by mwiebe** [comment]
Halide is definitely something I'm excited about, and in a similar vein ISPC (https://ispc.github.io/). Being able to define DyND callables with these systems would be awesome.

**Comment by insertinterestingnamehere** [comment]
We definitely are drawing on ideas from Julia. That said, one important thing to keep in mind here is that the DyND core isn't really meant to define a particular syntax. We're building an ndarray-oriented type system and dispatch system that can be embedded in a variety of languages through a core ABI. Currently our C++ library supplies both the core functionality and additional syntax for interacting with it in C++. Our Python bindings define the way we interact with the DyND core in Python. Hopefully we'll be able to interact with more languages and syntaxes at some point as well.

**Comment by datnamer** [comment]
So the sort of "traits" that I mentioned earlier is part of an open [issue](https://github.com/python/typing/issues/11) in the python typing package, seeming supported by no less than BDFL himself.

They are being called protocols or structural subtyping. I felt this is a  great  middle ground between totally open ducktyping and a rigid totally inheritance based type hierarchy. This way different types can simply implement interfaces without worrying about inheriting from a specific type that a functions expects. 

There was discussion on the mailing list of having this go into subsequent PEPs: https://mail.python.org/pipermail/python-dev/2015-April/139267.html

**Comment by datnamer** [comment]
Hi, 

There is an interesting and relevant (atleast for low level datashape maybe) discussion from an intel engineer here: https://groups.google.com/forum/#!topic/python-ideas/WNZQysbFbnw 

"n the future (and the future is now) application programs will 
want to be aware of what class of memory they are allocating:  normal DRAM, 
persistent RAM, fast DRAM, (etc?).  High level languages like Python that 
manage memory for the application programmer will need to expose a way 
for programmers to declare which kind of memory backs an object, and 
define rules for interactions between objects that reside in different 
memory types.  I'm interested in starting a discussion (not necessarily 
here, though I think that probably makes sense) about the challenges 
this presents and how Python-the-language and CPython the implementation 
might solve them. "

Wasn't sure where to put that and didn't want to open a new issue.

**Comment by insertinterestingnamehere** [comment]
So, at a really high level, different storage types and layouts fit really well as DyND types. Our computation kernels already provide some of the basic building blocks necessary for computations that can be run independent of the storage layout. This is mostly already possible and is something that DyND is already designed to handle in the long term. On the other hand, we're currently working a lot more on building better handling of in-memory computations, better core infrastructure, and better user-facing interfaces.

**Comment by datnamer** [comment]
"better handling of in-memory computations"

Isn't by definition Dynd handling in memory use cases? I thought Dask would be a layer over Dynd that would scale to distributed and out of core. 

The problem is though, that Dask is pure python so I'm not sure how it would work with Dynd's multi dispatch type lattice.  (like I want a custom OOC type and define operations on it, but then it can't participate in Dynd's dispatch machinery and thus generic kernels).

**Comment by insertinterestingnamehere** [comment]
Right, the issues you mentioned with interop with other libraries are all currently open questions. DyND would make a great backend for a lot of the stuff in Dask. We've wanted DyND to support defining types in languages other than C++ in an extensible way for a long time, and that would help a lot with interoperability between DyND and these other packages. Even just with the types and layouts we already support, the functionality DyND already has could theoretically be valuable as an in-memory backend for Dask. On the other hand, the DyND type system is already designed to be able to handle alternative memory layouts and storage locations entirely on its own. It's really a matter of what gets implemented.

**Comment by datnamer** [comment]
Interesting.

What exactly would interfacing with python objects entail? Wouldn't Dynd need some sort of concrete typed memory guarantee and semantic restrictions to do so?

Then we are getting into the syntactical and technical complexity of almost another python compiler with its own type system...so we have Numba with its JIT class annotations, datashape and then Mypy. 

Numba and others are also considering how  to work with this issue: https://python-compilers-workshop.github.io/

Perhaps it would be good to solve this problem generally (provide a way for compilers to easily reason about python code), and then have Dynd utilize this path. 

Since Mypy already seems to be getting some momentum, has better and simpler syntax (imo) and already has type inference and some older compiler hooks, what would you think about standardizing on either just the syntax or even the infrastructure as a central type system between all these projects? Then datashape can be a special case.  This way Dynd can boostrap off all the cool type system stuff happening like protocols etc

This would maybe provide a path to  mypy style  optional static compile time typing..which would make python unique among up and coming technical computing languages.

Of course, this could all a pipe dream so feel free to knock this idea down to earth.
